File: engine/src/testing/runner.ts
/**
 * Mock-execution runner: interprets a flow definition against a {@link TestCase}
 * and evaluates its assertions. Connector / child-flow / HTTP actions are never
 * called — their outputs come from the case's `mocks`, or default to an empty
 * successful result (and are reported in {@link TestResult.unmockedExternal}).
 *
 * Supported control flow: `runAfter` ordering + status gating, Scope, Condition
 * (`If`), Foreach, Switch, Terminate, and the variable/Compose actions.
 * Documented limits: `Until` runs its body once; parallel branches are executed
 * in a deterministic order; expression coverage is the subset in `expr.ts`.
 */
import type { FlowModel } from '../types.js';
import type { ActionStatus, ActionTrace, Assertion, AssertionResult, MockAction, TestCase, TestResult } from './types.js';
import { emptyContext, resolveValue, evaluateExpression, toBool } from './expr.js';
import type { EvalContext } from './expr.js';

function asRecord(v: unknown): Record<string, unknown> {
  return v && typeof v === 'object' && !Array.isArray(v) ? (v as Record<string, unknown>) : {};
}
const asStrArr = (v: unknown): string[] => (Array.isArray(v) ? v.map(String) : []);
const toNum = (v: unknown): number => (typeof v === 'number' ? v : Number(v));

interface State {
  ctx: EvalContext;
  trace: ActionTrace[];
  statuses: Record<string, ActionStatus>;
  branches: Record<string, 'yes' | 'no'>;
  mocks: Record<string, MockAction>;
  unmocked: Set<string>;
  terminated: { status: ActionStatus } | null;
}

const EXTERNAL_TYPES = new Set([
  'openapiconnection', 'openapiconnectionwebhook', 'openapiconnectionnotification',
  'apiconnection', 'apiconnectionwebhook', 'http', 'https', 'workflow', 'function',
  'sendemail', 'sendemailv2', 'response',
]);

export function runTestCase(model: FlowModel, testCase: TestCase): TestResult {
  const def = asRecord(model.raw);
  const ctx = emptyContext();
  ctx.trigger.outputs = testCase.trigger?.outputs;
  ctx.trigger.body = testCase.trigger?.body ?? asRecord(testCase.trigger?.outputs).body;
  if (testCase.variables) ctx.variables = { ...testCase.variables };

  const state: State = { ctx, trace: [], statuses: {}, branches: {}, mocks: testCase.mocks ?? {}, unmocked: new Set(), terminated: null };
  runScope(asRecord(def.actions), state);

  const assertions = testCase.asserts.map((a) => evalAssertion(a, state));
  return {
    caseName: testCase.name,
    passed: assertions.every((x) => x.passed),
    assertions,
    trace: state.trace,
    unmockedExternal: [...state.unmocked],
  };
}

interface ScopeOutcome { status: ActionStatus; children: Array<{ name: string; status: ActionStatus; outputs?: unknown }>; }

function runScope(actionsMap: Record<string, unknown>, state: State): ScopeOutcome {
  const entries = Object.entries(actionsMap);
  const local: Record<string, ActionStatus> = {};
  const done = new Set<string>();
  const children: ScopeOutcome['children'] = [];
  let progress = true;
  while (done.size < entries.length && progress) {
    progress = false;
    for (const [name, raw] of entries) {
      if (done.has(name)) continue;
      const action = asRecord(raw);
      const runAfter = asRecord(action.runAfter);
      const deps = Object.keys(runAfter);
      if (!deps.every((d) => done.has(d))) continue; // wait for predecessors

      let status: ActionStatus;
      const gated = deps.some((d) => !asStrArr(runAfter[d]).includes(local[d] ?? 'Skipped'));
      if (state.terminated || gated) {
        status = 'Skipped';
        pushTrace(state, { name, type: String(action.type ?? 'Unknown'), status });
        markSkipped(action, state); // cascade Skipped into nested actions
      } else {
        status = executeAction(name, action, state);
      }
      local[name] = status;
      state.statuses[name] = status;
      done.add(name);
      children.push({ name, status, outputs: state.ctx.outputs[name] });
      progress = true;
    }
  }
  const anyFailed = entries.some(([n]) => local[n] === 'Failed' || local[n] === 'TimedOut');
  return { status: anyFailed ? 'Failed' : 'Succeeded', children };
}

function executeAction(name: string, action: Record<string, unknown>, state: State): ActionStatus {
  const type = String(action.type ?? 'Unknown').toLowerCase();
  const mock = state.mocks[name];

  if (mock) {
    const status: ActionStatus = mock.status ?? (mock.error ? 'Failed' : 'Succeeded');
    setOutput(name, mock.outputs, state);
    pushTrace(state, { name, type: String(action.type), status, output: mock.outputs, mocked: true, note: mock.error });
    return status;
  }

  switch (type) {
    case 'compose': {
      const out = resolveValue(action.inputs, state.ctx);
      setOutput(name, out, state);
      pushTrace(state, { name, type: String(action.type), status: 'Succeeded', output: out });
      return 'Succeeded';
    }
    case 'initializevariable': {
      for (const v of asArray(asRecord(action.inputs).variables)) {
        const vv = asRecord(v);
        state.ctx.variables[String(vv.name)] = vv.value === undefined ? defaultForType(vv.type) : resolveValue(vv.value, state.ctx);
      }
      pushTrace(state, { name, type: String(action.type), status: 'Succeeded' });
      return 'Succeeded';
    }
    case 'setvariable': {
      const inp = asRecord(action.inputs);
      state.ctx.variables[String(inp.name)] = resolveValue(inp.value, state.ctx);
      pushTrace(state, { name, type: String(action.type), status: 'Succeeded' });
      return 'Succeeded';
    }
    case 'incrementvariable':
    case 'decrementvariable': {
      const inp = asRecord(action.inputs);
      const cur = toNum(state.ctx.variables[String(inp.name)] ?? 0);
      const by = toNum(resolveValue(inp.value ?? 1, state.ctx));
      state.ctx.variables[String(inp.name)] = type === 'incrementvariable' ? cur + by : cur - by;
      pushTrace(state, { name, type: String(action.type), status: 'Succeeded' });
      return 'Succeeded';
    }
    case 'appendtoarrayvariable':
    case 'appendtostringvariable': {
      const inp = asRecord(action.inputs);
      const key = String(inp.name);
      const val = resolveValue(inp.value, state.ctx);
      if (type === 'appendtoarrayvariable') {
        const arr = Array.isArray(state.ctx.variables[key]) ? (state.ctx.variables[key] as unknown[]) : [];
        arr.push(val); state.ctx.variables[key] = arr;
      } else {
        state.ctx.variables[key] = String(state.ctx.variables[key] ?? '') + String(val);
      }
      pushTrace(state, { name, type: String(action.type), status: 'Succeeded' });
      return 'Succeeded';
    }
    case 'scope': {
      const outcome = runScope(asRecord(action.actions), state);
      state.ctx.results[name] = outcome.children.map((c) => ({ name: c.name, status: c.status, outputs: c.outputs }));
      pushTrace(state, { name, type: String(action.type), status: outcome.status });
      return outcome.status;
    }
    case 'if': {
      const yes = evalCondition(action.expression, state.ctx);
      const branch: 'yes' | 'no' = yes ? 'yes' : 'no';
      state.branches[name] = branch;
      const body = yes ? asRecord(action.actions) : asRecord(asRecord(action.else).actions);
      pushTrace(state, { name, type: String(action.type), status: 'Succeeded', branch });
      const outcome = runScope(body, state);
      state.statuses[name] = outcome.status;
      return outcome.status;
    }
    case 'foreach': {
      const arr = resolveValue(action.foreach, state.ctx);
      const items = Array.isArray(arr) ? arr : [];
      let status: ActionStatus = 'Succeeded';
      for (const it of items) {
        state.ctx.currentItem = it; state.ctx.items[name] = it;
        const outcome = runScope(asRecord(action.actions), state);
        if (outcome.status === 'Failed') status = 'Failed';
      }
      pushTrace(state, { name, type: String(action.type), status });
      return status;
    }
    case 'switch': {
      const val = resolveValue(action.expression, state.ctx);
      const cases = asRecord(action.cases);
      let branch = asRecord(asRecord(action.default).actions);
      for (const c of Object.values(cases)) {
        const cr = asRecord(c);
        if (String(resolveValue(cr.case, state.ctx)) === String(val)) { branch = asRecord(cr.actions); break; }
      }
      pushTrace(state, { name, type: String(action.type), status: 'Succeeded' });
      return runScope(branch, state).status;
    }
    case 'terminate': {
      const runStatus = String(asRecord(action.inputs).runStatus ?? 'Cancelled');
      const status: ActionStatus = runStatus === 'Failed' ? 'Failed' : 'Succeeded';
      state.terminated = { status };
      pushTrace(state, { name, type: String(action.type), status });
      return status;
    }
    default: {
      if (EXTERNAL_TYPES.has(type)) {
        state.unmocked.add(name);
        setOutput(name, {}, state);
        pushTrace(state, { name, type: String(action.type), status: 'Succeeded', output: {}, note: 'external action assumed Succeeded — provide a mock for real coverage' });
        return 'Succeeded';
      }
      pushTrace(state, { name, type: String(action.type), status: 'Succeeded' });
      return 'Succeeded';
    }
  }
}

/* structured OR string condition */
function evalCondition(expr: unknown, ctx: EvalContext): boolean {
  if (typeof expr === 'string') return toBool(resolveValue(expr, ctx));
  const obj = asRecord(expr);
  const key = Object.keys(obj)[0];
  if (!key) return false;
  const raw = obj[key];
  const list = Array.isArray(raw) ? raw : [raw];
  switch (key) {
    case 'and': return list.every((e) => evalCondition(e, ctx));
    case 'or': return list.some((e) => evalCondition(e, ctx));
    case 'not': return !evalCondition(list[0], ctx);
    default: {
      const a = resolveValue(list[0], ctx); const b = resolveValue(list[1], ctx);
      return compare(key, a, b);
    }
  }
}
function compare(op: string, a: unknown, b: unknown): boolean {
  switch (op) {
    case 'equals': return JSON.stringify(a) === JSON.stringify(b);
    case 'notequals': return JSON.stringify(a) !== JSON.stringify(b);
    case 'greater': return toNum(a) > toNum(b);
    case 'greaterorequals': return toNum(a) >= toNum(b);
    case 'less': return toNum(a) < toNum(b);
    case 'lessorequals': return toNum(a) <= toNum(b);
    case 'contains': return Array.isArray(a) ? a.some((x) => JSON.stringify(x) === JSON.stringify(b)) : String(a).includes(String(b));
    case 'startswith': return String(a).startsWith(String(b));
    case 'endswith': return String(a).endsWith(String(b));
    default: return false;
  }
}

/* mark every nested action of a skipped block as Skipped so assertions see it */
function markSkipped(action: Record<string, unknown>, state: State): void {
  const nested: unknown[] = [action.actions, asRecord(action.else).actions, asRecord(action.default).actions];
  for (const c of Object.values(asRecord(action.cases))) nested.push(asRecord(c).actions);
  for (const n of nested) {
    for (const [name, raw] of Object.entries(asRecord(n))) {
      state.statuses[name] = 'Skipped';
      pushTrace(state, { name, type: String(asRecord(raw).type ?? 'Unknown'), status: 'Skipped' });
      markSkipped(asRecord(raw), state);
    }
  }
}

function setOutput(name: string, outputs: unknown, state: State): void {
  state.ctx.outputs[name] = outputs;
  state.ctx.bodies[name] = outputs && typeof outputs === 'object' && 'body' in (outputs as object)
    ? (outputs as Record<string, unknown>).body : outputs;
}
function pushTrace(state: State, t: ActionTrace): void { state.trace.push(t); }
function asArray(v: unknown): unknown[] { return Array.isArray(v) ? v : v == null ? [] : [v]; }
function defaultForType(t: unknown): unknown {
  switch (String(t).toLowerCase()) {
    case 'array': return []; case 'object': return {}; case 'integer': case 'float': return 0;
    case 'boolean': return false; default: return '';
  }
}

/* ------------------------------ assertions ------------------------------ */
function evalAssertion(a: Assertion, state: State): AssertionResult {
  const ran = (n: string) => { const s = state.statuses[n]; return s !== undefined && s !== 'Skipped'; };
  const ok = (passed: boolean, message: string): AssertionResult => ({ assertion: a, passed, message });
  switch (a.type) {
    case 'ran': return ok(ran(a.action), ran(a.action) ? `${a.action} ran` : `${a.action} did not run (status ${state.statuses[a.action] ?? 'absent'})`);
    case 'skipped': { const sk = (state.statuses[a.action] ?? 'Skipped') === 'Skipped'; return ok(sk, sk ? `${a.action} was skipped` : `${a.action} ran (status ${state.statuses[a.action]})`); }
    case 'status': { const s = state.statuses[a.action]; return ok(s === a.equals, `${a.action} status = ${s ?? 'absent'} (expected ${a.equals})`); }
    case 'branchTaken': { const b = state.branches[a.condition]; return ok(b === a.branch, `${a.condition} took '${b ?? 'none'}' (expected '${a.branch}')`); }
    case 'variableEquals': { const v = state.ctx.variables[a.name]; const p = JSON.stringify(v) === JSON.stringify(a.equals); return ok(p, `variables('${a.name}') = ${JSON.stringify(v)} (expected ${JSON.stringify(a.equals)})`); }
    case 'outputEquals':
    case 'expression': {
      let val: unknown; let err = '';
      try { const e = a.expression.startsWith('@') ? a.expression.slice(1) : a.expression; val = evaluateExpression(e, state.ctx); }
      catch (x) { err = (x as Error).message; }
      if (err) return ok(false, `expression '${a.expression}' failed: ${err}`);
      const p = JSON.stringify(val) === JSON.stringify(a.equals);
      return ok(p, `${a.expression} = ${JSON.stringify(val)} (expected ${JSON.stringify(a.equals)})`);
    }
    case 'noFailures': {
      const failed = Object.entries(state.statuses).filter(([, s]) => s === 'Failed' || s === 'TimedOut').map(([n]) => n);
      return ok(failed.length === 0, failed.length ? `actions failed: ${failed.join(', ')}` : 'no actions failed');
    }
  }
}

File: plugin/Ember.Plugins/Testing/Runner.cs
using System;
using System.Collections.Generic;
using System.Linq;
using BetaRover.Ember.Engine;
using BetaRover.Ember.Json;

namespace BetaRover.Ember.Testing
{
    /// <summary>
    /// Port of engine/src/testing/runner.ts — interprets a flow definition against a
    /// <see cref="TestCase"/> and evaluates its assertions. Connector / child-flow /
    /// HTTP actions are never called; their outputs come from the case's mocks, or
    /// default to an empty successful result (reported in
    /// <see cref="TestResult.UnmockedExternal"/>).
    ///
    /// Supported control flow: runAfter ordering + status gating, Scope, Condition
    /// (If), Foreach, Switch, Terminate, and the variable/Compose actions. Documented
    /// limits: Until runs its body once; parallel branches run in definition order;
    /// expression coverage is the subset in <see cref="Expr"/>.
    /// </summary>
    public static class Runner
    {
        private const string Succeeded = "Succeeded";
        private const string Failed = "Failed";
        private const string Skipped = "Skipped";
        private const string TimedOut = "TimedOut";

        private static readonly HashSet<string> ExternalTypes = new HashSet<string>
        {
            "openapiconnection", "openapiconnectionwebhook", "openapiconnectionnotification",
            "apiconnection", "apiconnectionwebhook", "http", "https", "workflow", "function",
            "sendemail", "sendemailv2", "response",
        };

        private sealed class State
        {
            public EvalContext Ctx;
            public List<ActionTrace> Trace = new List<ActionTrace>();
            public Dictionary<string, string> Statuses = new Dictionary<string, string>();
            public Dictionary<string, string> Branches = new Dictionary<string, string>();
            public Dictionary<string, MockAction> Mocks;
            public HashSet<string> Unmocked = new HashSet<string>();
            public string Terminated; // null, or the terminal status
        }

        private static Dictionary<string, object> AsRecord(object v) => JsonUtil.AsObject(v);
        private static List<object> AsArray(object v)
        {
            if (v is List<object> l) return l;
            if (v == null) return new List<object>();
            return new List<object> { v };
        }
        private static double ToNum(object v) => Expr.ToNum(v);

        /// <summary>Parse a flow's clientdata and run a batch of cases against it.</summary>
        public static List<TestResult> RunSuite(string displayName, object clientData, List<TestCase> cases)
        {
            var model = FlowParser.ParseFlow(displayName, clientData);
            return cases.Select(c => RunTestCase(model, c)).ToList();
        }

        public static TestResult RunTestCase(FlowModel model, TestCase testCase)
        {
            var def = AsRecord(model.Raw);
            var ctx = EvalContext.Empty();
            ctx.TriggerOutputs = testCase.TriggerOutputs;
            ctx.TriggerBody = testCase.HasTriggerBody ? testCase.TriggerBody : JsonUtil.Get(AsRecord(testCase.TriggerOutputs), "body");
            if (testCase.Variables != null) ctx.Variables = new Dictionary<string, object>(testCase.Variables);

            var state = new State { Ctx = ctx, Mocks = testCase.Mocks ?? new Dictionary<string, MockAction>() };
            RunScope(AsRecord(JsonUtil.Get(def, "actions")), state);

            var assertions = testCase.Asserts.Select(a => EvalAssertion(a, state)).ToList();
            return new TestResult
            {
                CaseName = testCase.Name,
                Passed = assertions.All(x => x.Passed),
                Assertions = assertions,
                Trace = state.Trace,
                UnmockedExternal = state.Unmocked.ToList(),
            };
        }

        private sealed class ScopeChild { public string Name; public string Status; public object Outputs; }
        private sealed class ScopeOutcome { public string Status; public List<ScopeChild> Children; }

        private static ScopeOutcome RunScope(Dictionary<string, object> actionsMap, State state)
        {
            var entries = actionsMap.ToList();
            var local = new Dictionary<string, string>();
            var done = new HashSet<string>();
            var children = new List<ScopeChild>();
            bool progress = true;
            while (done.Count < entries.Count && progress)
            {
                progress = false;
                foreach (var entry in entries)
                {
                    string name = entry.Key;
                    if (done.Contains(name)) continue;
                    var action = AsRecord(entry.Value);
                    var runAfter = AsRecord(JsonUtil.Get(action, "runAfter"));
                    var deps = runAfter.Keys.ToList();
                    if (!deps.All(d => done.Contains(d))) continue; // wait for predecessors

                    string status;
                    bool gated = deps.Any(d =>
                    {
                        var allowed = AsArray(JsonUtil.Get(runAfter, d)).Select(Expr.ToStr).ToList();
                        string prior = local.ContainsKey(d) ? local[d] : Skipped;
                        return !allowed.Contains(prior);
                    });
                    if (state.Terminated != null || gated)
                    {
                        status = Skipped;
                        PushTrace(state, new ActionTrace { Name = name, Type = Expr.ToStr(JsonUtil.Get(action, "type") ?? "Unknown"), Status = status });
                        MarkSkipped(action, state); // cascade Skipped into nested actions
                    }
                    else
                    {
                        status = ExecuteAction(name, action, state);
                    }
                    local[name] = status;
                    state.Statuses[name] = status;
                    done.Add(name);
                    object outp; state.Ctx.Outputs.TryGetValue(name, out outp);
                    children.Add(new ScopeChild { Name = name, Status = status, Outputs = outp });
                    progress = true;
                }
            }
            bool anyFailed = entries.Any(e => local.ContainsKey(e.Key) && (local[e.Key] == Failed || local[e.Key] == TimedOut));
            return new ScopeOutcome { Status = anyFailed ? Failed : Succeeded, Children = children };
        }

        private static string ExecuteAction(string name, Dictionary<string, object> action, State state)
        {
            string typeRaw = Expr.ToStr(JsonUtil.Get(action, "type") ?? "Unknown");
            string type = typeRaw.ToLowerInvariant();
            MockAction mock;
            if (state.Mocks.TryGetValue(name, out mock) && mock != null)
            {
                string mstatus = mock.Status ?? (mock.Error != null ? Failed : Succeeded);
                SetOutput(name, mock.Outputs, state);
                PushTrace(state, new ActionTrace { Name = name, Type = typeRaw, Status = mstatus, Output = mock.Outputs, HasOutput = true, Mocked = true, Note = mock.Error });
                return mstatus;
            }

            switch (type)
            {
                case "compose":
                {
                    var outp = Expr.ResolveValue(JsonUtil.Get(action, "inputs"), state.Ctx);
                    SetOutput(name, outp, state);
                    PushTrace(state, new ActionTrace { Name = name, Type = typeRaw, Status = Succeeded, Output = outp, HasOutput = true });
                    return Succeeded;
                }
                case "initializevariable":
                {
                    var inputs = AsRecord(JsonUtil.Get(action, "inputs"));
                    foreach (var v in AsArray(JsonUtil.Get(inputs, "variables")))
                    {
                        var vv = AsRecord(v);
                        object val = vv.ContainsKey("value") ? Expr.ResolveValue(vv["value"], state.Ctx) : DefaultForType(JsonUtil.Get(vv, "type"));
                        state.Ctx.Variables[Expr.ToStr(JsonUtil.Get(vv, "name"))] = val;
                    }
                    PushTrace(state, new ActionTrace { Name = name, Type = typeRaw, Status = Succeeded });
                    return Succeeded;
                }
                case "setvariable":
                {
                    var inp = AsRecord(JsonUtil.Get(action, "inputs"));
                    state.Ctx.Variables[Expr.ToStr(JsonUtil.Get(inp, "name"))] = Expr.ResolveValue(JsonUtil.Get(inp, "value"), state.Ctx);
                    PushTrace(state, new ActionTrace { Name = name, Type = typeRaw, Status = Succeeded });
                    return Succeeded;
                }
                case "incrementvariable":
                case "decrementvariable":
                {
                    var inp = AsRecord(JsonUtil.Get(action, "inputs"));
                    string key = Expr.ToStr(JsonUtil.Get(inp, "name"));
                    object curRaw; state.Ctx.Variables.TryGetValue(key, out curRaw);
                    double cur = ToNum(curRaw ?? (double)0);
                    object byRaw = inp.ContainsKey("value") ? inp["value"] : (double)1;
                    double by = ToNum(Expr.ResolveValue(byRaw, state.Ctx));
                    state.Ctx.Variables[key] = type == "incrementvariable" ? cur + by : cur - by;
                    PushTrace(state, new ActionTrace { Name = name, Type = typeRaw, Status = Succeeded });
                    return Succeeded;
                }
                case "appendtoarrayvariable":
                case "appendtostringvariable":
                {
                    var inp = AsRecord(JsonUtil.Get(action, "inputs"));
                    string key = Expr.ToStr(JsonUtil.Get(inp, "name"));
                    object val = Expr.ResolveValue(JsonUtil.Get(inp, "value"), state.Ctx);
                    if (type == "appendtoarrayvariable")
                    {
                        object cur; state.Ctx.Variables.TryGetValue(key, out cur);
                        var arr = cur as List<object> ?? new List<object>();
                        arr.Add(val); state.Ctx.Variables[key] = arr;
                    }
                    else
                    {
                        object cur; state.Ctx.Variables.TryGetValue(key, out cur);
                        state.Ctx.Variables[key] = Expr.ToStr(cur ?? "") + Expr.ToStr(val);
                    }
                    PushTrace(state, new ActionTrace { Name = name, Type = typeRaw, Status = Succeeded });
                    return Succeeded;
                }
                case "scope":
                {
                    var outcome = RunScope(AsRecord(JsonUtil.Get(action, "actions")), state);
                    state.Ctx.Results[name] = outcome.Children
                        .Select(c => (object)new Dictionary<string, object> { { "name", c.Name }, { "status", c.Status }, { "outputs", c.Outputs } })
                        .ToList();
                    PushTrace(state, new ActionTrace { Name = name, Type = typeRaw, Status = outcome.Status });
                    return outcome.Status;
                }
                case "if":
                {
                    bool yes = EvalCondition(JsonUtil.Get(action, "expression"), state.Ctx);
                    string branch = yes ? "yes" : "no";
                    state.Branches[name] = branch;
                    var body = yes ? AsRecord(JsonUtil.Get(action, "actions")) : AsRecord(JsonUtil.Get(AsRecord(JsonUtil.Get(action, "else")), "actions"));
                    PushTrace(state, new ActionTrace { Name = name, Type = typeRaw, Status = Succeeded, Branch = branch });
                    var outcome = RunScope(body, state);
                    state.Statuses[name] = outcome.Status;
                    return outcome.Status;
                }
                case "foreach":
                {
                    var arr = Expr.ResolveValue(JsonUtil.Get(action, "foreach"), state.Ctx);
                    var items = arr as List<object> ?? new List<object>();
                    string status = Succeeded;
                    foreach (var it in items)
                    {
                        state.Ctx.CurrentItem = it; state.Ctx.Items[name] = it;
                        var outcome = RunScope(AsRecord(JsonUtil.Get(action, "actions")), state);
                        if (outcome.Status == Failed) status = Failed;
                    }
                    PushTrace(state, new ActionTrace { Name = name, Type = typeRaw, Status = status });
                    return status;
                }
                case "switch":
                {
                    var val = Expr.ResolveValue(JsonUtil.Get(action, "expression"), state.Ctx);
                    var cases = AsRecord(JsonUtil.Get(action, "cases"));
                    var branch = AsRecord(JsonUtil.Get(AsRecord(JsonUtil.Get(action, "default")), "actions"));
                    foreach (var c in cases.Values)
                    {
                        var cr = AsRecord(c);
                        if (Expr.ToStr(Expr.ResolveValue(JsonUtil.Get(cr, "case"), state.Ctx)) == Expr.ToStr(val))
                        { branch = AsRecord(JsonUtil.Get(cr, "actions")); break; }
                    }
                    PushTrace(state, new ActionTrace { Name = name, Type = typeRaw, Status = Succeeded });
                    return RunScope(branch, state).Status;
                }
                case "terminate":
                {
                    string runStatus = Expr.ToStr(JsonUtil.Get(AsRecord(JsonUtil.Get(action, "inputs")), "runStatus") ?? "Cancelled");
                    string status = runStatus == "Failed" ? Failed : Succeeded;
                    state.Terminated = status;
                    PushTrace(state, new ActionTrace { Name = name, Type = typeRaw, Status = status });
                    return status;
                }
                default:
                {
                    if (ExternalTypes.Contains(type))
                    {
                        state.Unmocked.Add(name);
                        SetOutput(name, new Dictionary<string, object>(), state);
                        PushTrace(state, new ActionTrace { Name = name, Type = typeRaw, Status = Succeeded, Output = new Dictionary<string, object>(), HasOutput = true, Note = "external action assumed Succeeded — provide a mock for real coverage" });
                        return Succeeded;
                    }
                    PushTrace(state, new ActionTrace { Name = name, Type = typeRaw, Status = Succeeded });
                    return Succeeded;
                }
            }
        }

        /* structured OR string condition */
        private static bool EvalCondition(object expr, EvalContext ctx)
        {
            if (expr is string s) return Expr.ToBool(Expr.ResolveValue(s, ctx));
            var obj = AsRecord(expr);
            if (obj.Count == 0) return false;
            var key = obj.Keys.First();
            var raw = obj[key];
            var list = raw is List<object> l ? l : new List<object> { raw };
            switch (key)
            {
                case "and": return list.All(e => EvalCondition(e, ctx));
                case "or": return list.Any(e => EvalCondition(e, ctx));
                case "not": return !EvalCondition(list.Count > 0 ? list[0] : null, ctx);
                default:
                {
                    object a = Expr.ResolveValue(list.Count > 0 ? list[0] : null, ctx);
                    object b = Expr.ResolveValue(list.Count > 1 ? list[1] : null, ctx);
                    return Compare(key, a, b);
                }
            }
        }

        private static bool Compare(string op, object a, object b)
        {
            switch (op)
            {
                case "equals": return JsonUtil.Serialize(a) == JsonUtil.Serialize(b);
                case "notequals": return JsonUtil.Serialize(a) != JsonUtil.Serialize(b);
                case "greater": return ToNum(a) > ToNum(b);
                case "greaterorequals": return ToNum(a) >= ToNum(b);
                case "less": return ToNum(a) < ToNum(b);
                case "lessorequals": return ToNum(a) <= ToNum(b);
                case "contains": return a is List<object> l ? l.Any(x => JsonUtil.Serialize(x) == JsonUtil.Serialize(b)) : Expr.ToStr(a).IndexOf(Expr.ToStr(b), StringComparison.Ordinal) >= 0;
                case "startswith": return Expr.ToStr(a).StartsWith(Expr.ToStr(b), StringComparison.Ordinal);
                case "endswith": return Expr.ToStr(a).EndsWith(Expr.ToStr(b), StringComparison.Ordinal);
                default: return false;
            }
        }

        /* mark every nested action of a skipped block as Skipped so assertions see it */
        private static void MarkSkipped(Dictionary<string, object> action, State state)
        {
            var nested = new List<object>
            {
                JsonUtil.Get(action, "actions"),
                JsonUtil.Get(AsRecord(JsonUtil.Get(action, "else")), "actions"),
                JsonUtil.Get(AsRecord(JsonUtil.Get(action, "default")), "actions"),
            };
            foreach (var c in AsRecord(JsonUtil.Get(action, "cases")).Values)
                nested.Add(JsonUtil.Get(AsRecord(c), "actions"));
            foreach (var nAct in nested)
            {
                foreach (var kv in AsRecord(nAct))
                {
                    state.Statuses[kv.Key] = Skipped;
                    PushTrace(state, new ActionTrace { Name = kv.Key, Type = Expr.ToStr(JsonUtil.Get(AsRecord(kv.Value), "type") ?? "Unknown"), Status = Skipped });
                    MarkSkipped(AsRecord(kv.Value), state);
                }
            }
        }

        private static void SetOutput(string name, object outputs, State state)
        {
            state.Ctx.Outputs[name] = outputs;
            if (outputs is Dictionary<string, object> o && o.ContainsKey("body"))
                state.Ctx.Bodies[name] = o["body"];
            else
                state.Ctx.Bodies[name] = outputs;
        }

        private static void PushTrace(State state, ActionTrace t) => state.Trace.Add(t);

        private static object DefaultForType(object t)
        {
            switch (Expr.ToStr(t).ToLowerInvariant())
            {
                case "array": return new List<object>();
                case "object": return new Dictionary<string, object>();
                case "integer":
                case "float": return (double)0;
                case "boolean": return false;
                default: return "";
            }
        }

        /* ------------------------------ assertions ------------------------------ */
        private static AssertionResult EvalAssertion(Assertion a, State state)
        {
            Func<string, bool> ran = n => { string s; return state.Statuses.TryGetValue(n, out s) && s != Skipped; };
            Func<bool, string, AssertionResult> ok = (passed, message) => new AssertionResult { Assertion = a, Passed = passed, Message = message };
            string s0;
            switch (a.Type)
            {
                case "ran":
                    return ok(ran(a.Action), ran(a.Action) ? a.Action + " ran" : a.Action + " did not run (status " + (state.Statuses.TryGetValue(a.Action, out s0) ? s0 : "absent") + ")");
                case "skipped":
                {
                    string s = state.Statuses.TryGetValue(a.Action, out s0) ? s0 : Skipped;
                    bool sk = s == Skipped;
                    return ok(sk, sk ? a.Action + " was skipped" : a.Action + " ran (status " + s + ")");
                }
                case "status":
                {
                    string s = state.Statuses.TryGetValue(a.Action, out s0) ? s0 : null;
                    return ok(s == Expr.ToStr(a.Expected), a.Action + " status = " + (s ?? "absent") + " (expected " + Expr.ToStr(a.Expected) + ")");
                }
                case "branchTaken":
                {
                    string b = state.Branches.TryGetValue(a.Condition, out s0) ? s0 : null;
                    return ok(b == a.Branch, a.Condition + " took '" + (b ?? "none") + "' (expected '" + a.Branch + "')");
                }
                case "variableEquals":
                {
                    object v; state.Ctx.Variables.TryGetValue(a.Name, out v);
                    bool p = JsonUtil.Serialize(v) == JsonUtil.Serialize(a.Expected);
                    return ok(p, "variables('" + a.Name + "') = " + JsonUtil.Serialize(v) + " (expected " + JsonUtil.Serialize(a.Expected) + ")");
                }
                case "outputEquals":
                case "expression":
                {
                    object val = null; string err = null;
                    try { string e = a.Expression.StartsWith("@") ? a.Expression.Substring(1) : a.Expression; val = Expr.EvaluateExpression(e, state.Ctx); }
                    catch (Exception x) { err = x.Message; }
                    if (err != null) return ok(false, "expression '" + a.Expression + "' failed: " + err);
                    bool p = JsonUtil.Serialize(val) == JsonUtil.Serialize(a.Expected);
                    return ok(p, a.Expression + " = " + JsonUtil.Serialize(val) + " (expected " + JsonUtil.Serialize(a.Expected) + ")");
                }
                case "noFailures":
                {
                    var failed = state.Statuses.Where(kv => kv.Value == Failed || kv.Value == TimedOut).Select(kv => kv.Key).ToList();
                    return ok(failed.Count == 0, failed.Count > 0 ? "actions failed: " + string.Join(", ", failed) : "no actions failed");
                }
                default:
                    return ok(false, "unknown assertion type '" + a.Type + "'");
            }
        }
    }
}
